rustworkx-core/src/coloring.rs

```diff
@@ -141,21 +176,190 @@ where
+        if let Some(color) = preset_color_fn(k)? {
+            colors.insert(k, color);
+            for v in graph.neighbors(k) {
+                nbd_colors.get_mut(&v).unwrap().insert(color);
+            }
+        }
+    }
+
+    // Add non-preset nodes to priority queue
+    for k in graph.node_identifiers() {
+        if colors.get(&k).is_none() {
+            let degree = graph
+                .neighbors(k)
+                .filter(|v| colors.get(v).is_none())
+                .count();
+            let saturation = nbd_colors.get(&k).unwrap().len();
+            pq.push(k, SaturationStrategyData { degree, saturation });
+        }
+    }
+
+    // Greedily process nodes
+    while let Some((k, _)) = pq.pop() {
+        let neighbor_colors = nbd_colors.get(&k).unwrap();
+        let mut current_color: usize = 0;
+        loop {
+            if !neighbor_colors.contains(&current_color) {
+                break;
+            }
+            current_color += 1;
+        }
+
+        colors.insert(k, current_color);
+        for v in graph.neighbors(k) {
+            if colors.get(&v).is_none() {
+                nbd_colors.get_mut(&v).unwrap().insert(current_color);
+                let (_, vdata) = pq.get(&v).unwrap();
+
+                pq.push(
+                    v,
+                    SaturationStrategyData {
+                        degree: vdata.degree - 1,
+                        saturation: nbd_colors.get(&v).unwrap().len(),
+                    },
+                );
+            }
+        }
+    }
+
+    Ok(colors)
+}
+
+fn inner_greedy_node_color_strategy_independent_set<G, F, E>(
+    graph: G,
+    mut preset_color_fn: F,
+) -> Result<DictMap<G::NodeId, usize>, E>
+where
+    G: NodeCount + IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::NodeId: Hash + Eq + Send + Sync,
+    F: FnMut(G::NodeId) -> Result<Option<usize>, E>,
+{
+    let mut colors: DictMap<G::NodeId, usize> = DictMap::with_capacity(graph.node_count());
+
+    let mut preset: HashSet<G::NodeId> = HashSet::new();
+    let mut unprocessed: IndexSet<G::NodeId, RandomState> =
+        IndexSet::with_hasher(RandomState::default());
+
+    // Handle preset nodes
+    for k in graph.node_identifiers() {
+        if let Some(color) = preset_color_fn(k)? {
+            colors.insert(k, color);
+            preset.insert(k);
+        } else {
+            unprocessed.insert(k);
+        }
+    }
+
+    let mut current_color = 0;
+    while !unprocessed.is_empty() {
+        let mut remaining: IndexSet<G::NodeId, RandomState> =
+            IndexSet::with_hasher(RandomState::default());
+
+        // Remove neighbors of preset nodes with the given color
+        for k in &preset {
+            if colors.get(k) == Some(&current_color) {
+                for u in graph.neighbors(*k) {
+                    if unprocessed.swap_take(&u).is_some() {
+                        remaining.insert(u);
+                    }
+                }
+            }
+        }
```

Comment:
I think you can use an iterator for this first loop which I think will be a bit faster because it should give compilers hints about the size of `remaining`. It's something like this I think:
(this is untested so I'm sure I made typos or other mistakes, but I think it's the basic idea)

```suggestion
        // Remove neighbors of preset nodes with the given color
        let mut remaining: IndexSet<G::NodeId, RandomState> = preset.iter().filter(|k| {
            colors.get(k) == Some(&current_color)
        })
        .flat_map(|k| {
                graph.neighbors(*k).filter(|u| {
                    unprocessed.swap_take(&u).is_some()
                }
        })
        .collect();
```

Reply:
This is the only suggestion that I was not able to incorporate. Thanks to this, I have learned about the `flat_map` and I understand the code intention, it's really clever how you chain everything together. However, after the following minor fixes:
```
let mut remaining: IndexSet<G::NodeId, RandomState> = preset
	.iter()
	.filter(|k| colors.get(*k) == Some(&current_color))
	.flat_map(|k| {
		graph
			.neighbors(*k)
			.filter(|u| unprocessed.swap_take(u).is_some())
	})
	.collect();
```
I am getting the following compilation error
```
error: captured variable cannot escape `FnMut` closure body                                                                                                                            
   --> rustworkx-core\src\coloring.rs:303:17
    |
284 |       let mut unprocessed: IndexSet<G::NodeId, RandomState> =
    |           --------------- variable defined here
...
302 |               .flat_map(|k| {
    |                           - inferred to be a `FnMut` closure
303 | /                 graph
304 | |                     .neighbors(*k)
305 | |                     .filter(|u| unprocessed.swap_take(u).is_some())
    | |_________________________________-----------_______________________^ returns a closure that contains a reference to a captured variable, which then escapes the closure body     
    |                                   |
    |                                   variable captured here
    |
    = note: `FnMut` closures only have access to their captured variables while they are executing...
    = note: ...therefore, they cannot allow references to captured variables to escape
help: consider adding 'move' keyword before the nested closure
    |
305 |                     .filter(move |u| unprocessed.swap_take(u).is_some())
    |                             ++++

error: could not compile `rustworkx-core` (lib) due to previous error                                                                                                                  
```
I think I kind of understand the error but I am not sure how to fix it, the automatic suggestion to use `move` leads to even more problems.

Reply:
Ah too bad, yeah I missed the ownership issue around mutation of `unprocessed`. Rust won't let us do that because it would require it can't track the mutable borrows with the closure. Let's just leave the loop then it's not worth fighting this when a loop works fine.